mainpy/__init__.py

```diff
@@ -93,7 +95,14 @@ def main(
     if not callable(function):
         raise TypeError(f'expected a callable, got {type(function).__name__}')
 
-    if function.__module__ != '__main__':
+    if hasattr(sys, '_getframe'):
+        frame: Optional[FrameType] = sys._getframe(1).f_back  # noqa
+        assert frame is not None
+        name = frame.f_globals['__name__']
+    else:
+        name = function.__module__
```

Comment:
performance-wise it might be best to check if `function.__module__ == '__main__` , before traversing the call stack